utils/logger.go

```diff
@@ -3,40 +3,51 @@ package utils
 import (
 	"io"
 	"os"
-	"strconv"
-	"sync"
+	"runtime"
 	"time"
 
 	"github.com/rs/zerolog"
 	"github.com/rs/zerolog/pkgerrors"
 )
 
-var once sync.Once
+const (
+	DEFAULT_LOG_LEVEL      = zerolog.DebugLevel
+	LOG_TIME_FORMAT_FULL   = time.RFC3339
+	LOG_CALLER_STACK_DEPTH = 3 // XXX YA: Hack-y
+)
+
+var logger zerolog.Logger
 
-var log zerolog.Logger
+func GetLogger() *zerolog.Logger {
+	return &logger
+}
 
-func GetLogger() zerolog.Logger {
-	once.Do(func() {
-		zerolog.ErrorStackMarshaler = pkgerrors.MarshalStack
-		zerolog.TimeFieldFormat = time.RFC3339Nano
+type LineInfoHook struct{}
+
+func (h LineInfoHook) Run(e *zerolog.Event, l zerolog.Level, msg string) {
+	_, file, line, ok := runtime.Caller(LOG_CALLER_STACK_DEPTH)
+	if ok && l >= zerolog.WarnLevel {
+		e.Str("file", file)
+		e.Int("line", line)
+	}
+}
 
-		logLevel, err := strconv.Atoi(os.Getenv("LOG_LEVEL"))
-		if err != nil {
-			logLevel = int(zerolog.DebugLevel)
-		}
+func init() {
+	zerolog.ErrorStackMarshaler = pkgerrors.MarshalStack
 
-		var output io.Writer = zerolog.ConsoleWriter{
-			Out:        os.Stdout,
-			TimeFormat: time.RFC3339,
-		}
+	logLevelStr := os.Getenv("CEDANA_LOG_LEVEL")
+	logLevel, err := zerolog.ParseLevel(logLevelStr)
+	if err != nil || logLevelStr == "" { // allow turning off logging
+		logLevel = DEFAULT_LOG_LEVEL
+	}
 
-		log = zerolog.New(output).
-			Level(zerolog.Level(logLevel)).
-			With().
-			Timestamp().
-			Logger().
-			Output(zerolog.ConsoleWriter{Out: os.Stdout})
-	})
+	var output io.Writer = zerolog.ConsoleWriter{
+		Out: os.Stdout,
+	}
 
-	return log
+	logger = zerolog.New(output).
+		Level(logLevel).
+		With().
+		Timestamp().
+		Logger().Hook(LineInfoHook{})
```

Comment:
Oh, great! But we don't want line numbers for each message do we? It would clutter. Let me check this out, if we can configure it for only warning/errors and above levels.

Reply:
Okay, `caller` is pretty concise. I changed it to use it when the currently set log level is debug or lower.